File: app/src/main/java/com/drdisagree/iconify/xposed/mods/DepthWallpaper.java
package com.drdisagree.iconify.xposed.mods;

import static com.drdisagree.iconify.common.Const.SYSTEMUI_PACKAGE;
import static com.drdisagree.iconify.common.Preferences.DEPTH_WALLPAPER_SWITCH;
import static com.drdisagree.iconify.common.Preferences.ICONIFY_DEPTH_WALLPAPER_TAG;
import static com.drdisagree.iconify.common.Preferences.UNZOOM_DEPTH_WALLPAPER;
import static com.drdisagree.iconify.config.XPrefs.Xprefs;
import static de.robv.android.xposed.XposedBridge.hookAllMethods;
import static de.robv.android.xposed.XposedHelpers.findClass;
import static de.robv.android.xposed.XposedHelpers.findClassIfExists;
import static de.robv.android.xposed.XposedHelpers.getObjectField;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.ImageDecoder;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.drdisagree.iconify.xposed.ModPack;

import java.io.File;
import java.util.Objects;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

@SuppressLint("DiscouragedApi")
public class DepthWallpaper extends ModPack {

    private static final String TAG = "Iconify - " + DepthWallpaper.class.getSimpleName() + ": ";
    private boolean showDepthWallpaper = false;
    private FrameLayout mDepthWallpaperLayout = null;
    private ImageView mDepthWallpaperBackground = null;
    private ImageView mDepthWallpaperForeground = null;

    public DepthWallpaper(Context context) {
        super(context);
    }

    @Override
    public void updatePrefs(String... Key) {
        if (Xprefs == null) return;

        showDepthWallpaper = Xprefs.getBoolean(DEPTH_WALLPAPER_SWITCH, false);

        if (Key.length > 0 && Objects.equals(Key[0], DEPTH_WALLPAPER_SWITCH)) {
            updateWallpaper();
        }
    }

    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam lpparam) {
        Class<?> KeyguardBottomAreaViewClass = findClass(SYSTEMUI_PACKAGE + ".statusbar.phone.KeyguardBottomAreaView", lpparam.classLoader);

        hookAllMethods(KeyguardBottomAreaViewClass, "onFinishInflate", new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) {
                if (!showDepthWallpaper) return;

                View view = (View) param.thisObject;
                ViewGroup mIndicationArea = view.findViewById(mContext.getResources().getIdentifier("keyguard_indication_area", "id", mContext.getPackageName()));

                mIndicationArea.setClipChildren(false);
                mIndicationArea.setClipToPadding(false);
                mIndicationArea.getLayoutParams().height = ViewGroup.LayoutParams.MATCH_PARENT;
                mIndicationArea.getLayoutParams().width = ViewGroup.LayoutParams.MATCH_PARENT;
                ((ViewGroup.MarginLayoutParams) mIndicationArea.getLayoutParams()).bottomMargin = 0;

                // Create a new layout for the indication text views
                LinearLayout mIndicationTextView = new LinearLayout(mContext);
                LinearLayout.LayoutParams mIndicationViewParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
                int bottomMargin = mContext.getResources().getDimensionPixelSize(mContext.getResources().getIdentifier("keyguard_indication_margin_bottom", "dimen", mContext.getPackageName()));
                mIndicationViewParams.setMargins(0, 0, 0, bottomMargin);
                mIndicationViewParams.gravity = Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL;
                mIndicationTextView.setGravity(Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL);
                mIndicationTextView.setOrientation(LinearLayout.VERTICAL);
                mIndicationTextView.setLayoutParams(mIndicationViewParams);

                // Add the indication text views to the new layout
                TextView mTopIndicationView = mIndicationArea.findViewById(mContext.getResources().getIdentifier("keyguard_indication_text", "id", mContext.getPackageName()));
                TextView mLockScreenIndicationView = mIndicationArea.findViewById(mContext.getResources().getIdentifier("keyguard_indication_text_bottom", "id", mContext.getPackageName()));

                // We added a blank view to the top of the layout to push the indication text views to the bottom
                // The reason we did this is because gravity is not working properly on the indication text views
                View blankView = new View(mContext);
                blankView.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1.0f));

                // Remove the existing indication text views from the indication area
                ((ViewGroup) mTopIndicationView.getParent()).removeView(mTopIndicationView);
                ((ViewGroup) mLockScreenIndicationView.getParent()).removeView(mLockScreenIndicationView);

                // Add the indication text views to the new layout
                mIndicationTextView.addView(blankView);
                mIndicationTextView.addView(mTopIndicationView);
                mIndicationTextView.addView(mLockScreenIndicationView);

                FrameLayout mIndicationAreaDupe = new FrameLayout(mContext);
                mIndicationAreaDupe.setLayoutParams(new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
                mIndicationAreaDupe.addView(mIndicationTextView, -1);
                mIndicationArea.addView(mIndicationAreaDupe);

                // Get the depth wallpaper layout
                mDepthWallpaperLayout = mIndicationArea.findViewWithTag(ICONIFY_DEPTH_WALLPAPER_TAG);

                // Create the depth wallpaper layout if it doesn't exist
                if (mDepthWallpaperLayout == null) {
                    mDepthWallpaperLayout = new FrameLayout(mContext);
                    mDepthWallpaperLayout.setLayoutParams(new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
                    mDepthWallpaperLayout.setTag(ICONIFY_DEPTH_WALLPAPER_TAG);
                    mIndicationAreaDupe.addView(mDepthWallpaperLayout, 0);
                }

                mDepthWallpaperBackground = new ImageView(mContext);
                mDepthWallpaperForeground = new ImageView(mContext);

                mDepthWallpaperBackground.setLayoutParams(new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
                mDepthWallpaperForeground.setLayoutParams(new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));

                mDepthWallpaperLayout.addView(mDepthWallpaperBackground, 0);
                mDepthWallpaperLayout.addView(mDepthWallpaperForeground, -1);

                // Fix the bottom shortcuts pushing the wallpaper
                int[] offset = {0};
                try {
                    offset[0] = mContext.getResources().getDimensionPixelSize(mContext.getResources().getIdentifier("keyguard_affordance_fixed_height", "dimen", mContext.getPackageName()))
                            + mContext.getResources().getDimensionPixelSize(mContext.getResources().getIdentifier("keyguard_affordance_horizontal_offset", "dimen", mContext.getPackageName()));
                } catch (Throwable ignored) {
                }

                try {
                    ImageView startButton = view.findViewById(mContext.getResources().getIdentifier("start_button", "id", mContext.getPackageName()));
                    startButton.getViewTreeObserver().addOnGlobalLayoutListener(() -> ((ViewGroup.MarginLayoutParams) mIndicationTextView.getLayoutParams()).setMarginStart(startButton.getVisibility() == View.VISIBLE ? offset[0] : 0));
                } catch (Throwable ignored) {
                }

                try {
                    ImageView endButton = view.findViewById(mContext.getResources().getIdentifier("end_button", "id", mContext.getPackageName()));
                    endButton.getViewTreeObserver().addOnGlobalLayoutListener(() -> ((ViewGroup.MarginLayoutParams) mIndicationTextView.getLayoutParams()).setMarginEnd(endButton.getVisibility() == View.VISIBLE ? offset[0] : 0));
                } catch (Throwable ignored) {
                }

                updateWallpaper();
            }
        });

        hookAllMethods(KeyguardBottomAreaViewClass, "onConfigurationChanged", new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) {
                updateWallpaper();
            }
        });

        Class<?> NotificationPanelViewControllerClass = findClassIfExists(SYSTEMUI_PACKAGE + ".shade.NotificationPanelViewController", lpparam.classLoader);
        if (NotificationPanelViewControllerClass == null)
            NotificationPanelViewControllerClass = findClass(SYSTEMUI_PACKAGE + ".statusbar.phone.NotificationPanelViewController", lpparam.classLoader);

        hookAllMethods(NotificationPanelViewControllerClass, "onFinishInflate", new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) {
                if (!showDepthWallpaper) return;

                View mView = (View) getObjectField(param.thisObject, "mView");
                View keyguardBottomArea = mView.findViewById(mContext.getResources().getIdentifier("keyguard_bottom_area", "id", mContext.getPackageName()));
                ViewGroup parent = (ViewGroup) keyguardBottomArea.getParent();
                parent.removeView(keyguardBottomArea);
                parent.addView(keyguardBottomArea, 0);
            }
        });

        XC_MethodHook noKeyguardIndicationPadding = new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) {
                if (showDepthWallpaper) {
                    if (Build.VERSION.SDK_INT >= 33) {
                        try {
                            int resId = mContext.getResources().getIdentifier("keyguard_indication_area_padding", "dimen", mContext.getPackageName());
                            if (param.args[0].equals(resId)) {
                                param.setResult(0);
                            }
                        } catch (Throwable ignored) {
                        }
                    } else {
                        // These resources are only available on Android 12L and below
                        try {
                            int resId = mContext.getResources().getIdentifier("keyguard_indication_margin_bottom", "dimen", mContext.getPackageName());
                            if (param.args[0].equals(resId)) {
                                param.setResult(0);
                            }
                        } catch (Throwable ignored) {
                        }

                        try {
                            int resId = mContext.getResources().getIdentifier("keyguard_indication_margin_bottom_fingerprint_in_display", "dimen", mContext.getPackageName());
                            if (param.args[0].equals(resId)) {
                                param.setResult(0);
                            }
                        } catch (Throwable ignored) {
                        }
                    }
                }
            }
        };

        hookAllMethods(Resources.class, "getDimensionPixelOffset", noKeyguardIndicationPadding);
        hookAllMethods(Resources.class, "getDimensionPixelSize", noKeyguardIndicationPadding);
    }

    private void updateWallpaper() {
        if (mDepthWallpaperLayout == null) return;

        if (!showDepthWallpaper) {
            mDepthWallpaperLayout.setVisibility(View.GONE);
            return;
        }

        try {
            ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
            executor.scheduleAtFixedRate(() -> {
                File Android = new File(Environment.getExternalStorageDirectory() + "/Android");

                if (Android.isDirectory()) {
                    new Handler(Looper.getMainLooper()).post(() -> {
                        try {
                            ImageDecoder.Source backgroundImg = ImageDecoder.createSource(new File(Environment.getExternalStorageDirectory() + "/.iconify_files/depth_wallpaper_bg.png"));
                            ImageDecoder.Source foregroundImg = ImageDecoder.createSource(new File(Environment.getExternalStorageDirectory() + "/.iconify_files/depth_wallpaper_fg.png"));

                            Drawable backgroundDrawable = ImageDecoder.decodeDrawable(backgroundImg);
                            Drawable foregroundDrawable = ImageDecoder.decodeDrawable(foregroundImg);

                            mDepthWallpaperBackground.setImageDrawable(backgroundDrawable);
                            mDepthWallpaperBackground.setClipToOutline(true);
                            mDepthWallpaperBackground.setScaleType(ImageView.ScaleType.CENTER_CROP);

                            boolean zoomWallpaper = !Xprefs.getBoolean(UNZOOM_DEPTH_WALLPAPER, false);

                            if (zoomWallpaper) {
                                mDepthWallpaperBackground.setScaleX(1.1f);
                                mDepthWallpaperBackground.setScaleY(1.1f);
                            }

                            mDepthWallpaperForeground.setImageDrawable(foregroundDrawable);
                            mDepthWallpaperForeground.setClipToOutline(true);
                            mDepthWallpaperForeground.setScaleType(ImageView.ScaleType.CENTER_CROP);

                            if (zoomWallpaper) {
                                mDepthWallpaperForeground.setScaleX(1.1f);
                                mDepthWallpaperForeground.setScaleY(1.1f);
                            }

                            mDepthWallpaperLayout.setVisibility(View.VISIBLE);
                        } catch (Throwable ignored) {
                        }
                    });

                    executor.shutdown();
                    executor.shutdownNow();
                }
            }, 0, 5, TimeUnit.SECONDS);
        } catch (Throwable ignored) {
        }
    }
}

File: app/src/main/java/com/drdisagree/iconify/xposed/modules/DepthWallpaper.kt
package com.drdisagree.iconify.xposed.modules

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Resources
import android.graphics.ImageDecoder
import android.graphics.drawable.AnimatedImageDrawable
import android.graphics.drawable.Drawable
import android.os.Build
import android.os.Environment
import android.os.Handler
import android.os.Looper
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.MarginLayoutParams
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.drdisagree.iconify.common.Const.SYSTEMUI_PACKAGE
import com.drdisagree.iconify.common.Preferences.DEPTH_WALLPAPER_BACKGROUND_MOVEMENT_MULTIPLIER
import com.drdisagree.iconify.common.Preferences.DEPTH_WALLPAPER_CHANGED
import com.drdisagree.iconify.common.Preferences.DEPTH_WALLPAPER_FADE_ANIMATION
import com.drdisagree.iconify.common.Preferences.DEPTH_WALLPAPER_FOREGROUND_MOVEMENT_MULTIPLIER
import com.drdisagree.iconify.common.Preferences.DEPTH_WALLPAPER_PARALLAX_EFFECT
import com.drdisagree.iconify.common.Preferences.DEPTH_WALLPAPER_SWITCH
import com.drdisagree.iconify.common.Preferences.ICONIFY_DEPTH_WALLPAPER_TAG
import com.drdisagree.iconify.common.Preferences.ICONIFY_LOCKSCREEN_CLOCK_TAG
import com.drdisagree.iconify.common.Preferences.UNZOOM_DEPTH_WALLPAPER
import com.drdisagree.iconify.config.XPrefs.Xprefs
import com.drdisagree.iconify.xposed.ModPack
import com.drdisagree.iconify.xposed.modules.utils.ParallaxImageView
import com.drdisagree.iconify.xposed.modules.utils.ViewHelper.toPx
import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XposedBridge.hookAllMethods
import de.robv.android.xposed.XposedHelpers.findClass
import de.robv.android.xposed.XposedHelpers.findClassIfExists
import de.robv.android.xposed.XposedHelpers.getObjectField
import de.robv.android.xposed.callbacks.XC_LoadPackage.LoadPackageParam
import java.io.File
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

@SuppressLint("DiscouragedApi")
class DepthWallpaper(context: Context?) : ModPack(context!!) {

    private var showDepthWallpaper = false
    private var showFadingAnimation = false
    private var enableParallaxEffect = false
    private var backgroundMovement = 1.0f
    private var foregroundMovement = 3.0f
    private var mDepthWallpaperLayout: FrameLayout? = null
    private var mDepthWallpaperBackground: ParallaxImageView? = null
    private var mDepthWallpaperForeground: ParallaxImageView? = null
    private var mDozing = false
    private var unzoomWallpaper = false

    override fun updatePrefs(vararg key: String) {
        if (Xprefs == null) return

        showDepthWallpaper = Xprefs!!.getBoolean(DEPTH_WALLPAPER_SWITCH, false)
        showFadingAnimation = Xprefs!!.getBoolean(DEPTH_WALLPAPER_FADE_ANIMATION, false)
        enableParallaxEffect = Xprefs!!.getBoolean(DEPTH_WALLPAPER_PARALLAX_EFFECT, false)
        backgroundMovement = Xprefs!!.getFloat(DEPTH_WALLPAPER_BACKGROUND_MOVEMENT_MULTIPLIER, 1.0f)
        foregroundMovement = Xprefs!!.getFloat(DEPTH_WALLPAPER_FOREGROUND_MOVEMENT_MULTIPLIER, 3.0f)
        unzoomWallpaper = Xprefs!!.getBoolean(UNZOOM_DEPTH_WALLPAPER, false)

        if (key.isNotEmpty() &&
            (key[0] == DEPTH_WALLPAPER_SWITCH ||
                    key[0] == DEPTH_WALLPAPER_CHANGED ||
                    key[0] == DEPTH_WALLPAPER_BACKGROUND_MOVEMENT_MULTIPLIER ||
                    key[0] == DEPTH_WALLPAPER_FOREGROUND_MOVEMENT_MULTIPLIER ||
                    key[0] == UNZOOM_DEPTH_WALLPAPER)
        ) {
            updateWallpaper()
        }

        if (key.isNotEmpty() && key[0] == DEPTH_WALLPAPER_PARALLAX_EFFECT) {
            if (enableParallaxEffect) {
                mDepthWallpaperBackground?.registerSensorListener()
                mDepthWallpaperForeground?.registerSensorListener()
            } else {
                mDepthWallpaperBackground?.unregisterSensorListener()
                mDepthWallpaperForeground?.unregisterSensorListener()
            }
        }
    }

    override fun handleLoadPackage(loadPackageParam: LoadPackageParam) {
        val keyguardBottomAreaViewClass = findClass(
            "$SYSTEMUI_PACKAGE.statusbar.phone.KeyguardBottomAreaView",
            loadPackageParam.classLoader
        )

        hookAllMethods(keyguardBottomAreaViewClass, "onFinishInflate", object : XC_MethodHook() {
            override fun afterHookedMethod(param: MethodHookParam) {
                if (!showDepthWallpaper) return

                val view = param.thisObject as View
                val mIndicationArea = view.findViewById<ViewGroup>(
                    mContext.resources.getIdentifier(
                        "keyguard_indication_area",
                        "id",
                        mContext.packageName
                    )
                )
                mIndicationArea.setClipChildren(false)
                mIndicationArea.clipToPadding = false
                mIndicationArea.layoutParams.height = ViewGroup.LayoutParams.MATCH_PARENT
                mIndicationArea.layoutParams.width = ViewGroup.LayoutParams.MATCH_PARENT
                (mIndicationArea.layoutParams as MarginLayoutParams).bottomMargin = 0

                // Create a new layout for the indication text views
                val mIndicationView = LinearLayout(mContext)
                val mIndicationViewParams = LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )

                val bottomMargin = mContext.resources.getDimensionPixelSize(
                    mContext.resources.getIdentifier(
                        "keyguard_indication_margin_bottom",
                        "dimen",
                        mContext.packageName
                    )
                )

                mIndicationViewParams.setMargins(0, 0, 0, bottomMargin)
                mIndicationViewParams.gravity = Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
                mIndicationView.gravity = Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
                mIndicationView.orientation = LinearLayout.VERTICAL
                mIndicationView.setLayoutParams(mIndicationViewParams)

                // Get the indication text views
                val mTopIndicationView = mIndicationArea.findViewById<TextView>(
                    mContext.resources.getIdentifier(
                        "keyguard_indication_text",
                        "id",
                        mContext.packageName
                    )
                )
                val mLockScreenIndicationView = mIndicationArea.findViewById<TextView>(
                    mContext.resources.getIdentifier(
                        "keyguard_indication_text_bottom",
                        "id",
                        mContext.packageName
                    )
                )

                // Some roms also have FaceUnlockImageView
                val mFaceIconView: ImageView? = mIndicationArea.findViewById(
                    mContext.resources.getIdentifier(
                        "face_unlock_icon",
                        "id",
                        mContext.packageName
                    )
                )

                // We added a blank view to the top of the layout to push the indication text views to the bottom
                // The reason we did this is because gravity is not working properly on the indication text views
                val blankView = View(mContext)
                blankView.setLayoutParams(
                    LinearLayout.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        0,
                        1.0f
                    )
                )

                // Remove the existing indication text views from the indication area
                (mTopIndicationView.parent as ViewGroup).removeView(mTopIndicationView)
                (mLockScreenIndicationView.parent as ViewGroup).removeView(mLockScreenIndicationView)

                // Take care of the FaceUnlockImageView
                mFaceIconView?.let {
                    val layoutParams = it.layoutParams as MarginLayoutParams
                    layoutParams.topMargin =
                        mContext.resources.getDimensionPixelSize(
                            mContext.resources.getIdentifier(
                                "status_bar_height",
                                "dimen",
                                mContext.packageName
                            )
                        ) + mContext.toPx(30)

                    it.layoutParams = layoutParams
                    (it.parent as ViewGroup).removeView(it)
                    mIndicationView.addView(it)
                }

                // Add the indication text views to the new layout
                mIndicationView.addView(blankView)
                mIndicationView.addView(mTopIndicationView)
                mIndicationView.addView(mLockScreenIndicationView)

                val mIndicationAreaDupe = FrameLayout(mContext)
                mIndicationAreaDupe.setLayoutParams(
                    FrameLayout.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.MATCH_PARENT
                    )
                )
                mIndicationAreaDupe.addView(mIndicationView, -1)
                mIndicationArea.addView(mIndicationAreaDupe)

                // Get the depth wallpaper layout
                mDepthWallpaperLayout = mIndicationArea.findViewWithTag(ICONIFY_DEPTH_WALLPAPER_TAG)

                // Create the depth wallpaper layout if it doesn't exist
                if (mDepthWallpaperLayout == null) {
                    mDepthWallpaperLayout = FrameLayout(mContext)
                    mDepthWallpaperLayout!!.setLayoutParams(
                        FrameLayout.LayoutParams(
                            ViewGroup.LayoutParams.MATCH_PARENT,
                            ViewGroup.LayoutParams.MATCH_PARENT
                        )
                    )
                    mDepthWallpaperLayout!!.tag = ICONIFY_DEPTH_WALLPAPER_TAG
                    mIndicationAreaDupe.addView(mDepthWallpaperLayout, 0)
                }

                mDepthWallpaperBackground = ParallaxImageView(mContext)
                mDepthWallpaperForeground = ParallaxImageView(mContext)

                mDepthWallpaperBackground!!.setLayoutParams(
                    FrameLayout.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.MATCH_PARENT
                    )
                )
                mDepthWallpaperForeground!!.setLayoutParams(
                    FrameLayout.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.MATCH_PARENT
                    )
                )

                mDepthWallpaperLayout!!.addView(mDepthWallpaperBackground, 0)
                mDepthWallpaperLayout!!.addView(mDepthWallpaperForeground, -1)

                // Fix the bottom shortcuts pushing the wallpaper
                val affordanceFixedWidth = mContext.resources.getDimensionPixelSize(
                    mContext.resources.getIdentifier(
                        "keyguard_affordance_fixed_width",
                        "dimen",
                        mContext.packageName
                    )
                )
                val affordanceHorizontalOffset = mContext.resources.getDimensionPixelSize(
                    mContext.resources.getIdentifier(
                        "keyguard_affordance_horizontal_offset",
                        "dimen",
                        mContext.packageName
                    )
                )

                val offset = affordanceFixedWidth + affordanceHorizontalOffset + mContext.toPx(16)

                var startButton: ImageView? = null
                var endButton: ImageView? = null

                try {
                    startButton = view.findViewById(
                        mContext.resources.getIdentifier(
                            "start_button",
                            "id",
                            mContext.packageName
                        )
                    )
                    startButton.getViewTreeObserver().addOnGlobalLayoutListener {
                        (mIndicationView.layoutParams as MarginLayoutParams).setMarginStart(
                            if (startButton.visibility != View.GONE) offset else 0
                        )
                    }
                } catch (ignored: Throwable) {
                }

                try {
                    endButton = view.findViewById(
                        mContext.resources.getIdentifier(
                            "end_button",
                            "id",
                            mContext.packageName
                        )
                    )
                    endButton.getViewTreeObserver().addOnGlobalLayoutListener {
                        (mIndicationView.layoutParams as MarginLayoutParams).setMarginEnd(
                            if (endButton.visibility != View.GONE) offset else 0
                        )
                    }
                } catch (ignored: Throwable) {
                }

                if (Build.VERSION.SDK_INT >= 34) {
                    try {
                        val keyguardSettingsButton = view.findViewById<LinearLayout>(
                            mContext.resources.getIdentifier(
                                "keyguard_settings_button",
                                "id",
                                mContext.packageName
                            )
                        )
                        keyguardSettingsButton.getViewTreeObserver().addOnGlobalLayoutListener {
                            var marginStart = affordanceHorizontalOffset
                            var marginEnd = affordanceHorizontalOffset

                            startButton?.let {
                                marginStart = if (it.visibility != View.GONE) {
                                    offset
                                } else {
                                    affordanceHorizontalOffset
                                }
                            }
                            endButton?.let {
                                marginEnd = if (it.visibility != View.GONE) {
                                    offset
                                } else {
                                    affordanceHorizontalOffset
                                }
                            }

                            (mIndicationView.layoutParams as MarginLayoutParams).setMarginStart(
                                if (keyguardSettingsButton.visibility != View.GONE) marginStart else 0
                            )
                            (mIndicationView.layoutParams as MarginLayoutParams).setMarginEnd(
                                if (keyguardSettingsButton.visibility != View.GONE) marginEnd else 0
                            )
                        }
                    } catch (ignored: Throwable) {
                    }
                }

                updateWallpaper()
            }
        })

        hookAllMethods(
            keyguardBottomAreaViewClass,
            "onConfigurationChanged",
            object : XC_MethodHook() {
                override fun afterHookedMethod(param: MethodHookParam) {
                    updateWallpaper()
                }
            })

        var notificationPanelViewControllerClass = findClassIfExists(
            "$SYSTEMUI_PACKAGE.shade.NotificationPanelViewController",
            loadPackageParam.classLoader
        )
        if (notificationPanelViewControllerClass == null) notificationPanelViewControllerClass =
            findClass(
                "$SYSTEMUI_PACKAGE.statusbar.phone.NotificationPanelViewController",
                loadPackageParam.classLoader
            )

        val moveKeyguardBottomArea: XC_MethodHook = object : XC_MethodHook() {
            override fun afterHookedMethod(param: MethodHookParam) {
                if (!showDepthWallpaper) return

                val mView = getObjectField(param.thisObject, "mView") as View
                val keyguardBottomArea = mView.findViewById<View>(
                    mContext.resources.getIdentifier(
                        "keyguard_bottom_area",
                        "id",
                        mContext.packageName
                    )
                )

                val parent = keyguardBottomArea.parent as ViewGroup
                parent.removeView(keyguardBottomArea)
                parent.addView(keyguardBottomArea, 0)
            }
        }

        hookAllMethods(
            notificationPanelViewControllerClass,
            "onFinishInflate",
            moveKeyguardBottomArea
        )
        hookAllMethods(
            notificationPanelViewControllerClass,
            "reInflateViews",
            moveKeyguardBottomArea
        )

        val noKeyguardIndicationPadding: XC_MethodHook = object : XC_MethodHook() {
            override fun beforeHookedMethod(param: MethodHookParam) {
                if (showDepthWallpaper) {
                    if (Build.VERSION.SDK_INT >= 33) {
                        try {
                            val resId = mContext.resources.getIdentifier(
                                "keyguard_indication_area_padding",
                                "dimen",
                                mContext.packageName
                            )

                            if (param.args[0] == resId) {
                                param.result = 0
                            }
                        } catch (ignored: Throwable) {
                        }
                    } else {
                        // These resources are only available on Android 12L and below
                        try {
                            val resId = mContext.resources.getIdentifier(
                                "keyguard_indication_margin_bottom",
                                "dimen",
                                mContext.packageName
                            )

                            if (param.args[0] == resId) {
                                param.result = 0
                            }
                        } catch (ignored: Throwable) {
                        }
                        try {
                            val resId = mContext.resources.getIdentifier(
                                "keyguard_indication_margin_bottom_fingerprint_in_display",
                                "dimen",
                                mContext.packageName
                            )

                            if (param.args[0] == resId) {
                                param.result = 0
                            }
                        } catch (ignored: Throwable) {
                        }
                    }
                }
            }
        }

        hookAllMethods(
            Resources::class.java,
            "getDimensionPixelOffset",
            noKeyguardIndicationPadding
        )
        hookAllMethods(
            Resources::class.java,
            "getDimensionPixelSize",
            noKeyguardIndicationPadding
        )

        val dozeScrimControllerClass = findClass(
            "$SYSTEMUI_PACKAGE.statusbar.phone.DozeScrimController",
            loadPackageParam.classLoader
        )

        hookAllMethods(dozeScrimControllerClass, "onDozingChanged", object : XC_MethodHook() {
            override fun afterHookedMethod(param: MethodHookParam) {
                Handler(Looper.getMainLooper()).post {
                    updateFadeAnimation(
                        isDozing = param.args[0] as Boolean
                    )
                }
            }
        })
    }

    private fun updateWallpaper() {
        if (mDepthWallpaperLayout == null) return

        if (!showDepthWallpaper) {
            mDepthWallpaperLayout!!.visibility = View.GONE
            return
        }

        try {
            val executor = Executors.newSingleThreadScheduledExecutor()
            executor.scheduleAtFixedRate({
                val androidDir =
                    File(Environment.getExternalStorageDirectory().toString() + "/Android")
                if (androidDir.isDirectory()) {
                    Handler(Looper.getMainLooper()).post {
                        try {
                            val backgroundImg = ImageDecoder.createSource(
                                File(
                                    Environment.getExternalStorageDirectory()
                                        .toString() + "/.iconify_files/depth_wallpaper_bg.png"
                                )
                            )
                            val foregroundImg = ImageDecoder.createSource(
                                File(
                                    Environment.getExternalStorageDirectory()
                                        .toString() + "/.iconify_files/depth_wallpaper_fg.png"
                                )
                            )

                            val backgroundDrawable = ImageDecoder.decodeDrawable(backgroundImg)
                            val foregroundDrawable = ImageDecoder.decodeDrawable(foregroundImg)

                            mDepthWallpaperBackground!!.loadImageOrGif(backgroundDrawable)
                            mDepthWallpaperBackground!!.setMovementMultiplier(backgroundMovement)

                            mDepthWallpaperForeground!!.loadImageOrGif(foregroundDrawable)
                            mDepthWallpaperForeground!!.setMovementMultiplier(foregroundMovement)

                            mDepthWallpaperLayout!!.visibility = View.VISIBLE
                        } catch (ignored: Throwable) {
                        }
                    }

                    executor.shutdown()
                    executor.shutdownNow()
                }
            }, 0, 5, TimeUnit.SECONDS)
        } catch (ignored: Throwable) {
        }
    }

    private fun updateFadeAnimation(isDozing: Boolean) {
        if (!showDepthWallpaper || mDozing == isDozing) return

        val clockView = mDepthWallpaperLayout!!.findViewWithTag<View>(ICONIFY_LOCKSCREEN_CLOCK_TAG)
        val animDuration: Long = 800
        val startDelay: Long = 3000

        mDozing = isDozing

        if (!isDozing) {
            mDepthWallpaperBackground?.let {
                if (it.alpha != 1f) {
                    if (showFadingAnimation) {
                        val animation = it.animation

                        if (!(animation != null && animation.hasStarted() && !animation.hasEnded())) {
                            it.clearAnimation()

                            it.animate()
                                .alpha(1f)
                                .setDuration(animDuration)
                                .start()
                        }
                    } else {
                        it.setAlpha(1f)
                    }
                }
            }

            mDepthWallpaperForeground?.let {
                if (it.alpha != 1f) {
                    if (showFadingAnimation) {
                        val animation = it.animation

                        if (!(animation != null && animation.hasStarted() && !animation.hasEnded())) {
                            it.clearAnimation()

                            it.animate()
                                .alpha(1f)
                                .setDuration(animDuration)
                                .start()
                        }
                    } else {
                        it.setAlpha(1f)
                    }
                }
            }

            clockView?.let {
                if (it.alpha != 1f) {
                    if (showFadingAnimation) {
                        val animation = it.animation

                        if (!(animation != null && animation.hasStarted() && !animation.hasEnded())) {
                            it.clearAnimation()

                            it.animate()
                                .alpha(1f)
                                .setDuration(animDuration)
                                .start()
                        }
                    } else {
                        it.setAlpha(1f)
                    }
                }
            }
        } else {
            mDepthWallpaperBackground?.let {
                if (it.alpha != 0f && showFadingAnimation) {
                    val animation = it.animation

                    if (!(animation != null && animation.hasStarted() && !animation.hasEnded())) {
                        it.clearAnimation()

                        Handler(Looper.getMainLooper()).postDelayed({
                            it.animate()
                                .alpha(0f)
                                .setDuration(animDuration)
                                .start()
                        }, startDelay)
                    }
                }
            }

            mDepthWallpaperForeground?.let {
                if (it.alpha != 0f && showFadingAnimation) {
                    val animation = it.animation

                    if (!(animation != null && animation.hasStarted() && !animation.hasEnded())) {
                        it.clearAnimation()

                        Handler(Looper.getMainLooper()).postDelayed({
                            it.animate()
                                .alpha(0f)
                                .setDuration(animDuration)
                                .start()
                        }, startDelay)
                    }
                }
            }

            clockView?.let {
                if (it.alpha != 0.7f && showFadingAnimation) {
                    val animation = it.animation

                    if (!(animation != null && animation.hasStarted() && !animation.hasEnded())) {
                        it.clearAnimation()

                        Handler(Looper.getMainLooper()).postDelayed({
                            it.animate()
                                .alpha(0.7f)
                                .setDuration(animDuration)
                                .start()
                        }, startDelay)
                    }
                }
            }
        }
    }

    private fun ImageView.loadImageOrGif(drawable: Drawable) {
        setImageDrawable(drawable)
        setClipToOutline(true)
        setScaleType(ImageView.ScaleType.CENTER_CROP)

        if (!unzoomWallpaper) {
            scaleX = 1.1f
            scaleY = 1.1f
        } else {
            scaleX = 1.0f
            scaleY = 1.0f
        }

        if (drawable is AnimatedImageDrawable) {
            drawable.start()
        }
    }

    companion object {
        private val TAG = "Iconify - ${DepthWallpaper::class.java.simpleName}: "
    }
}